### PR TITLE
[ci] Add GitHub Actions bot to merge PRs on demand

### PR DESCRIPTION
**Pull request by driazati**
This implements https://discuss.tvm.apache.org/t/rfc-allow-merging-via-pr-comments/12220. The bot can be invoked from a top-level review comment or via a regular PR comment. The text `@tvm-bot merge` anywhere in the body will trigger the bot. Right now it checks that the latest commit is reviewed and that all CI jobs that have run on that commit are successful. If it fails, it will leave a comment on the PR with the reason.

This is just a start and some features are left for followups:
* Various TODOs throughout the code
* "Scheduled" merges that happen once CI finishes
* Allowing committers to merge without getting a fresh review for changes after an approval

Test PRs
Description | Link
------ | -----
ci passed: no, then yes<br>approved: no, then yes  | https://github.com/driazati/tvm/pull/13
ci passed: yes<br>approved: yes, then PR is updated, then re-approved  | https://github.com/driazati/tvm/pull/17
commit with a message | https://github.com/driazati/tvm/pull/16
cancelled merge | https://github.com/driazati/tvm/pull/18

cc @areusch @Mousius @gromero @kparzysz-quic

### CODE REVIEW

**Comment by gromero** [comment]
@driazati Thanks!

An initial question: for both examples #13 and #17, it seems that the final commit messages that land were not taken from the PR title/body (are empty?), but that's planned tho (WIP)?

**Comment by driazati** [comment]
Ah I could have used better examples, the code strips out [the specific text](https://github.com/apache/tvm/pull/10833/files#diff-705ee6ea53b898b6cc498839556fa1e0287d7d0496025cc9518792a1c24da605R312) `Thanks for contributing to TVM! ...` since that is just for PR authors and shouldn't really be in the commit log:

**Comment by gromero** [comment]
hmm neat :)

File: tests/python/ci/sample_prs/pr10786-badci.json
Comment (areusch) [inline tests/python/ci/sample_prs/pr10786-badci.json:1]: what about one where ci-merge was missing (e.g. ci-head vs ci-merge)?

Comment (areusch) [inline tests/python/ci/sample_prs/pr10786-badci.json:1]: are these test cases briefly described anywhere? we could e.g. add a key describing them or to body

File: tests/python/ci/sample_prs/pr10786-oldreview.json
Comment (areusch) [inline tests/python/ci/sample_prs/pr10786-oldreview.json:117]: haha this is a bit subtle, want to make it super different just to be clear?

File: tests/scripts/github_mergebot.py
Comment (areusch) [inline tests/scripts/github_mergebot.py:367]: just wondering how one might be sure they could stop merging after they've sent comment to merge. is this more like rescinding perm to merge?

File: tests/scripts/github_mergebot.py
Comment (areusch) [inline tests/scripts/github_mergebot.py:159]: why do we have this flag and also `--testing-pr-json`?

**Review by areusch — COMMENTED** [review]
thanks @driazati couple comments

File: tests/python/ci/sample_prs/pr10786-badci.json
Comment (driazati) [reply tests/python/ci/sample_prs/pr10786-badci.json:1]: is there something more than the result that you were looking for? i.e. the actual usages here https://github.com/apache/tvm/pull/10833/files#diff-0dd95b1adca3c8ab9745a9be219a5683885a9710365a447ee290841b172747a6R72-R91

File: tests/scripts/github_mergebot.py
Comment (driazati) [reply tests/scripts/github_mergebot.py:159]: one is for testing, the other is for development, we can probably delete this at some point once we have some more consensus on this PR (i.e. `DEBUG` will fetch + cache automatically, `--testing-pr-json` relies on a file being on disk)

File: tests/scripts/github_mergebot.py
Comment (driazati) [reply tests/scripts/github_mergebot.py:367]: one thing about cancels is that they'd have to be almost immediate, with GHA queueing + script runtime it's not more than 30 seconds before a merge comment to the merge actually happening. That's what this does, basically the comments are all in a list and if the last thing in the list is a merge command it'll merge, otherwise nothing happens. If someone later comments another merge it'll get merged as normal.

One thing I probably should add is a filter for who can engage a merge, probably should be just the PR author + committers

**Comment by areusch** [comment]
this one ready for a review?

**Comment by driazati** [comment]
yup it's ready for review, let me know if there's any more test cases you'd want to see as well

File: .github/workflows/merge.yml
Comment (areusch) [inline .github/workflows/merge.yml:17]: just a reminder to uncomment

File: tests/python/ci/sample_prs/pr10786-badci.json
Comment (areusch) [reply tests/python/ci/sample_prs/pr10786-badci.json:1]: just thinking whether we should have `{"testcase_description": "Expected JSON when CI has failed"}` in here

File: tests/scripts/github_mergebot.py
Comment (areusch) [reply tests/scripts/github_mergebot.py:159]: ok, could you just document this env var? with a comment somewhere?

File: tests/scripts/github_mergebot.py
Comment (areusch) [inline tests/scripts/github_mergebot.py:395]: just following up here--did we want to add a filter for who can merge? i think that is a good idea.

**Review by areusch — COMMENTED** [review]
cool this is looking pretty good. a few more questions, but i think we're getting close here.

File: tests/scripts/github_mergebot.py
Comment (driazati) [reply tests/scripts/github_mergebot.py:395]: added along with a test

File: tests/scripts/github_mergebot.py
Comment (driazati) [reply tests/scripts/github_mergebot.py:159]: added to the tool `--help`

File: tests/python/ci/sample_prs/pr10786-badci.json
Comment (driazati) [reply tests/python/ci/sample_prs/pr10786-badci.json:1]: added a `detail` to each test case with some expanation

**Review by areusch — APPROVED** [review]
thanks @driazati !

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:376]:
```suggestion
            logging.info(f"Comment {comment} was from committer comment and is mergeable")
```

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:375]: Can we get a test case for this?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:371]: We need a test case for when it's the author asking for a merge with a committers review and without, that's our core use-case I think?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:477]: Please add a test case for this

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:500]: Text needs updating

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:437]: Should `NEUTRAL` come under the failed case given it won't have ran?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:468]: Can split the CI logic out into its own function to keep `merge_if_passed_checks` just checking the various statuses?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:413]: I don't think this is fully implemented yet?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:348]: Should we remove this for now? We're not checking the return just yet.

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:426]: Can you just add the lists?

```suggestion
        all_comments = sorted(
            self.raw["comments"]["nodes"] + self.reviews(),
            key=lambda comment: comment["updatedAt"]
        )
```

**Review by Mousius — CHANGES_REQUESTED** [review]

File: tests/scripts/github_mergebot.py
Comment (driazati) [reply tests/scripts/github_mergebot.py:413]: can you elaborate? if the last comment addressed to the bot doesn't say `merge` it won't detect that a merge was requested (so if it's `cancel` then there will be no merge)

**Comment by driazati** [comment]
@Mousius can you re-review?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [reply tests/scripts/github_mergebot.py:410]: Oh sorry, I missed that logic, it filters out the `None` cases and `"cancel"` would be top of the list - gotcha. So this string is essentially `"anything-but-merge"`, seems ok now I read it again 😸

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:427]: I don't think this comment is necessary anymore?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:450]: I don't think this comment is necessary anymore?

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:464]:
```suggestion
        all_ci_passed = len(failed_ci_jobs) == 0
        has_one_approval = False

        if not all_ci_passed:
            failed_jobs_msg = "\n".join(
                [f" * [{job['name']} (`{job['status']}`)]({job['url']})" for job in failed_ci_jobs]
            )
            self.comment(
                f"Cannot merge, these CI jobs are not successful on {self.head_oid()}:\n{failed_jobs_msg}"
            )
            return
```

File: tests/scripts/github_mergebot.py
Comment (Mousius) [inline tests/scripts/github_mergebot.py:336]: Doesn't this mean we'll have weird commits like:

```
Message


Co-authored-by: Author1


Co-authored-by: Author2


Co-authored-by: Author3
```

Can we do anything neater here?

**Review by Mousius — CHANGES_REQUESTED** [review]
Minor comments left @driazati, otherwise I think this is getting pretty close 😸

**Review by Mousius — APPROVED** [review]

**Comment by Mousius** [comment]
Apologies for the delay @driazati , this LGTM and I'm glad to see this work finally land 😸

**Pull request merged**
